ath79: add support for LibreRouter v1

spiccinini opened the pull request:
Hardware
--------
SOC:   QCA9558
RAM:   128M DDR2
Flash: 16MiB SPI-NOR
ETH:   QCA8337N: 2x 10/100/1000 PoE and PoE pass-through
WiFi2: QCA9558 (bgn) 2T2R
WiFi5: 2x mPCIE with AR9582 (an) 2T2R
BTN:   1x Reset
GPIO:  multiple GPIO on header, PoE passthrough enable
UART:  3.3V 115200 8N1 header on the board
WDG:   ATTiny13 watchdog
JTAG:  header on the board
USB:   1x connector and 1x header on the board
PoE:   10-32V input in ETH port 1, passthrough in port 2
mPCIE: 2x populated with radios (but replaceable)

OpenWrt is preinstalled from factory.
Bootloader: https://github.com/librerouterorg/u-boot
Board details (specs, schematics, gerbers): https://github.com/librerouterorg/board


Signed-off-by: Santiago Piccinini <spiccinini@altermundi.net>

Based on this branch https://git.openwrt.org/?p=openwrt/staging/dangole.git;a=commit;h=6325eeec2d72a4b14883c46b8423ba8580cbd2e6 from Daniel Golle

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "u-boot";
```
Please make your u-boot partition read-only;

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+			};
+
+			partition@40000 {
+				label = "u-boot-env";
```
Since you are not using the uboot-envtools make this partition read-only.

spiccinini replied:
Yes the u-boot-env partition is used for dual-boot support (not included in this PR), it should be stated somewhere?

chunkeey replied:
Do you think you please could include it?

spiccinini replied:
Dual boot support for this device needs `CONFIG_MIPS_CMDLINE_DTB_EXTEND=y` or `CONFIG_MIPS_CMDLINE_FROM_BOOTLOADER=y`. Do you know if there is a way to enable this options only for this device? Beacause  AFAIK the other ath79 devices may break changing this config.

chunkeey replied:
> Dual boot support for this device needs CONFIG_MIPS_CMDLINE_DTB_EXTEND=y or CONFIG_MIPS_CMDLINE_FROM_BOOTLOADER=y. Do you know if there is a way to enable this options only for this device? Beacause AFAIK the other ath79 devices may break changing this config.

No, not with the MIPS target. However, it's possible to do that with PowerPC (yeah, this is a bit silly).

Now, Ideally you would want to support FIT so you could pass the dtb to the bootloader to make the changes to the dtb. However, you bootloader source is too old for this trick. However, it seems to support the multi-image, which can also be used to pass a dtb to the bootloader to be edited. What I don't know is if the ath79 MIPS linux kernel init code can deal with having the bootloader passing on the dtb.

spiccinini replied:
I will investigate FIT and multi-image, thanks! Dual boot support is not mandatory. The device is perfectly usable without using the dual-boot support and the images generated work fine. 

We may maintain out of tree support for dual-booting as also it is implemented using a the userspace tool `safe-upgrade` (https://github.com/libremesh/lime-packages/blob/master/packages/safe-upgrade/files/usr/sbin/safe-upgrade) that is not included in openwrt. dual-booting feature can be installed/bootstraped by the user using `safe-upgrade` so from factory the openwrt images will work. If the device has the dual-boot installed then installing images without cmdline support from bootloader won't work but the device won't brick (as it has automatic rollback).

spiccinini replied:
I am marking this as resolved because I don't think we can do more at least at this moment. Please reopen if you think otherwise

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+				reg = <0xfd0000 0x20000>;
+			};
+
+			art: partition@ff0000 {
```
Please make your ART partition read-only

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+	#size-cells = <0>;
+	status = "okay";
+
+	hub_port0: port@1 {
```
Just curious: Is there a plan to use this for a usbport LED trigger or something?

spiccinini replied:
Not from my side. It's better to remove the hub_port if not used? something like
```
        status = "okay";
-
-       hub_port0: port@1 {
-               reg = <1>;
-               #trigger-source-cells = <0>;
-       };
 };
```

chunkeey replied:
Have you tested them by repurposing one of the existing LEDs in the DTS? If this works, I would definitely leave them. But if you don't know and didn't test, you might was well remove them.

spiccinini replied:
No, I didn't test repurposing one of the exising LEDs for this.

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+
+	};
+
+	gpio-export {
```
Would it be possible to replace the gpio-export construct with the userspace gpio-switch equivalent?

(If you need a example device, please look at: fa3c2676ab0ce5fbb648249da31e31bbc61a3e3d )

spiccinini replied:
Why is gpio-switch prefered over gpio-export? 
Do you know if there is a "named  way" to use the gpio with gpio-switch? or to use it it's necesary to first do something like `uci get system.poe_passthrough.gpio_pin` to know the number of the gpio and then use it through `/sys/class/gpio/gpio1/`?

spiccinini replied:
I did the requested change, but still want to know why it is prefered :)

chunkeey replied:
> Why is gpio-switch prefered over gpio-export?

I think @mkresin said it best in his post:
https://github.com/openwrt/openwrt/pull/1366#issuecomment-424834448

But this topic comes up on the ML as well.

spiccinini replied:
Thanks!!

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+#include "qca9557.dtsi"
+
+/ {
+	compatible = "librerouter,librerouter-v1", "qca,qca9557";
```
If it's really a QCA9558, then I think "qca,qca9558" would be better here.

chunkeey commented on target/linux/ath79/base-files/etc/board.d/02_network:
```diff
@@ -116,6 +116,11 @@ ath79_setup_interfaces()
 			"0@eth0" "1:lan" "2:lan" "3:lan" "4:lan" "5:wan"
 	;;
+	librerouter,librerouter-v1)
+		ucidef_set_interfaces_lan_wan "eth1.2" "eth0.1"
```
```ucidef_set_interfaces_lan_wan "eth1.2" "eth0.1"```

Is this line still necessary? I was told some time ago that ucidef_add_switch will do the right thing here.

(Maybe the parameters for the switch0 just need to be reordered, Can you please check. Thanks)

spiccinini replied:
No, this line is not necesarry, thanks!

chunkeey reviewed (commented):
```OpenWrt is preinstalled from factory.```

Maybe it would still be a good idea to mention to use "sysupgrade -n" for the first install.
Also it doesn't hurt to have a word or two about tftpboot or another recovery method in
case the unit got bricked for some reason.

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+		compatible = "gpio-keys";
+
+		reset {
+			label = "Reset button";
```
I think "Reset" would be enough here. But you can (if you want) rename the "keys" parent node above to let's say "button" instead.

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+};
+
+&pcie0 {
+	status = "okay";
```
Could you please add the pci child nodes for the attached AR9582?

(I think the qca9558_tplink_tl-wdr4900-v2.dts could be a good example. It has a AR9580 but the IDs should be similar)

(Also make sure that the reg values are good if you are unsure please provide a bootlog)

spiccinini replied:
The owner of the board can change one or both of the m-pcie boards, let's say to use a 3g-modem, to change the functionality of the device. 
If we add the pci child nodes a new dts file will be neded ? What do you think it's the best way to handle this?

chunkeey replied:
Do you know if the AR9582 are preinstalled when the device "leaves the factory"? If yes, then I would include the nodes. 

Note: The pci DT enumeration works differently and the pcie subsystem code is able to handle cases where the pci device was replaced by something else.

spiccinini replied:
Yes, they are preinstalled from the factory.  Ok I didn't know that the pcie subsystem can handle replacements, I will test it ;)

spiccinini replied:
@chunkeey As far as I understand calibration and mac for this mpci cards should be read from EEPROM. Searching how to do that I found this commit https://github.com/torvalds/linux/commit/ce938231bd3b1d7af3cbd8836f084801090470e1 so If I understand correctly not using `qca,no-eeprom` is broken in 4.14?

dmesg extract:
```
[    0.000000] Linux version 4.14.95 (san@repo) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r0+9138-09b5402)) #0 Sun Jan 27 14:56:16 2019
...
[    0.189522] PCI host bridge to bus 0000:00
[    0.193887] pci_bus 0000:00: root bus resource [mem 0x10000000-0x13ffffff]
[    0.201234] pci_bus 0000:00: root bus resource [io  0x0000]
[    0.207155] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.214378] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.222828] pci 0000:00:00.0: [168c:0033] type 00 class 0x028000
[    0.222881] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x0001ffff 64bit]
[    0.222935] pci 0000:00:00.0: reg 0x30: [mem 0x00000000-0x0000ffff pref]
[    0.223009] pci 0000:00:00.0: supports D1
[    0.223017] pci 0000:00:00.0: PME# supported from D0 D1 D3hot
[    0.223220] pci 0000:00:00.0: BAR 0: assigned [mem 0x10000000-0x1001ffff 64bit]
[    0.231039] pci 0000:00:00.0: BAR 6: assigned [mem 0x10020000-0x1002ffff pref] 
...
[   11.002141] PCI: Enabling device 0000:00:00.0 (0000 -> 0002)             
[   11.099047] ath: EEPROM regdomain: 0x0                                   
[   11.099054] ath: EEPROM indicates default country code should be used    
[   11.099057] ath: doing EEPROM country->regdmn map search                 
[   11.099070] ath: country maps to regdmn code: 0x3a                       
[   11.099075] ath: Country alpha2 being used: US                           
[   11.099078] ath: Regpair used: 0x3a                                      
[   11.109284] ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[   11.110968] ieee80211 phy1: Atheros AR9300 Rev:4 mem=0xb0000000, irq=18  
```

chunkeey replied:
> I understand correctly not using qca,no-eeprom is broken in 4.14?

That issue is fixed in Openwrt's mac80211 package. As for the vanilla 4.14, yes this would be a problem there... But I did tell stable@vger.kernel.org to add that patch to a upcoming 4.14 stable release.

spiccinini replied:
Great, it works! :)

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+	status = "okay";
+};
+
+&pcie1 {
```
Same as with pcie0

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <25000000>;
+
```
If this is a SPI-NOR flash (the binding says so), you should definitely try setting the FAST READ property

"m25p,fast-read;"

for a cheap boost. (Please also run some tests).

spiccinini replied:
Great! didn't know about fast-read, will try and come back with results.

981213 replied:
With 25MHz SPI clock m25p,fast-read only slow-down the read operation. Fast read instructions is used for SPI clock higher than 50MHz and the extra dummy byte required by that instruction is a waste of 8 clocks when normal read instruction works fine.
There won't be any differences in ath79 with 4.14 kernel though, since we have a hacky flash_read function to use the memory-mapped area for reading operation.

spiccinini replied:
Thanks both, I will leave it without change for now.

chunkeey replied:
Ahh, 

> Great! didn't know about fast-read, will try and come back with results.

no results. I want to know, even if they are worse.

spiccinini replied:
Tested with 
```
                compatible = "jedec,spi-nor";
                reg = <0>;
                spi-max-frequency = <25000000>;
+               m25p,fast-read;
 
                partitions {
```
No improvement in boot time nor reading from /dev/mtd:
```
# 250000 MHz no fast-read
[   11.747702] kmodloader: done loading kernel modules from /etc/modules.d/*
root@OpenWrt:/# time cat /dev/mtd2 > /dev/null 
real    0m 5.45s
user    0m 0.00s
sys     0m 5.32s
root@OpenWrt:/# time cat /dev/mtd2 > /dev/null 
real    0m 5.45s
user    0m 0.00s
sys     0m 5.30s

# 25000 MHz fast-read
[   11.735793] kmodloader: done loading kernel modules from /etc/modules.d/*
root@OpenWrt:/# time cat /dev/mtd2 > /dev/null 
real    0m 5.45s
user    0m 0.00s
sys     0m 5.34s
root@OpenWrt:/# time cat /dev/mtd2 > /dev/null 
real    0m 5.44s
user    0m 0.01s
sys     0m 5.32s
```

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+			};
+
+			partition@fd0000 {
+				label = "res";
```
reserved? If so, for what? If you are interested you could utilize mtdoops to store panic logs in this partition.

spiccinini replied:
We thought about using one part of this area be used as a second copy of the u-boot-env (but still not enabled in the bootloader). We just left some space for new ideas, didn't know about mtdoops sounds interesting!

chunkeey commented on target/linux/ath79/dts/qca9558_librerouter-v1.dts:
```diff
@@ -0,0 +1,213 @@
+		compatible = "gpio-leds";
+
+		system: system {
+			label = "librerouter:green:system";
```
librerouter-v1:green:system ?
(This would match OpenWrt's convention of using the second part of the machine compatible as the board name)

spiccinini commented:
@chunkeey Thank you very much for the review! That was fast! I will do the fixes asap (not until past the weekend)

spiccinini committed:
ath79: add support for LibreRouter v1

Hardware
--------
SOC:   QCA9558
RAM:   128M DDR2
Flash: 16MiB SPI-NOR
ETH:   QCA8337N: 2x 10/100/1000 PoE and PoE pass-through
WiFi2: QCA9558 (bgn) 2T2R
WiFi5: 2x mPCIE with AR9582 (an) 2T2R
BTN:   1x Reset
GPIO:  multiple GPIO on header, PoE passthrough enable
UART:  3.3V 115200 8N1 header on the board
WDG:   ATTiny13 watchdog
JTAG:  header on the board
USB:   1x connector and 1x header on the board
PoE:   10-32V input in ETH port 1, passthrough in port 2
mPCIE: 2x populated with radios (but replaceable)

OpenWrt is preinstalled from factory. To install use <your-image>-sysupgade.bin
using the web interface or with sysupgrade -n.

Flash from bootloader (in case failsafe does not work)
1. Connect the LibreRouter with a serial adapter (TTL voltage) to the UART
   header in the board.
2. Connect an ETH cable and configure static ip addres 192.168.1.10/24
3. Turn on the device and stop the bootloader sending any key through the serial
   interface.
4. Use a TFTP server to serve <your image>-sysupgrade.bin file.
5. Execute the following commands at the bootloader prompt:
    ath> tftp 82000000 <your image>-sysupgrade.bin
    ath> erase 0x9f050000 +$filesize
    ath> cp.b 0x82000000 0x9f050000 $filesize
    ath> bootm 0x9f050000

More docs
* Bootloader https://github.com/librerouterorg/u-boot
* Board details (schematics, gerbers): https://github.com/librerouterorg/board

Signed-off-by: Santiago Piccinini <spiccinini@altermundi.net>

spiccinini commented:
Changes so far from the first commit:
- added mpci wifi nodes
- renamed `art` partition to `ART` fro ease dual-boot as this is the name used by the bootloader
- removed USB gpio ports
- fixed wan port switch config
- changed ART and u-boot partitions to read-only
- moved poe-passthrough dt `gpio-export` to `ucidef_add_gpio_switch` 
- changed system LED from `heartbeat` to `default on`
- added a blue LED for general purpose use
- renamed "Reset button" to "Reset"

All the changes where tested

chunkeey commented:
> removed USB gpio ports
USB LED trigger definitions?

Merged. Thanks.

adrianschmutzler commented:
Although this PR has been has been closed already, I would be interested in some additional data about the device(s) in case you still have access to it:

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.

neheb commented:
@spiccinini is this really using QCA8337 on QCA955x?

spiccinini commented:
> @spiccinini is this really using QCA8337 on QCA955x?

Yes!
